cmd/mt-kafka-mdm-report-out-of-order/inputooofinder.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/grafana/metrictank/idx/cassandra"
+	"github.com/grafana/metrictank/schema"
+	"github.com/grafana/metrictank/schema/msg"
+	log "github.com/sirupsen/logrus"
+)
+
+type Track struct {
+	Name string
+	Tags []string
+
+	LatestTimestamp int64
+}
+
+type Tracker map[schema.MKey]Track
+
+// find out of order metrics
+type inputOOOFinder struct {
+	prefix string
+	substr string
+
+	tracker                  Tracker
+	reorderBufferGracePeriod int64
+
+	groupByName   bool
+	groupedByName *map[string]int
+	groupByTag    string
+	groupedByTag  *map[string]int
+
+	lock sync.Mutex
+}
+
+func newInputOOOFinder(prefix string, substr string, partitionFrom int, partitionTo int, metricInterval int, reorderWindow int, groupByName bool, groupedByName *map[string]int, groupByTag string, groupedByTag *map[string]int) *inputOOOFinder {
+	cassandraIndex := cassandra.New(cassandra.CliConfig)
+	err := cassandraIndex.InitBare()
+	if err != nil {
+		log.Fatalf("error initializing cassandra index: %s", err.Error())
+		os.Exit(1)
+	}
+
+	metricDefinitions := make([]schema.MetricDefinition, 0)
+	for partition := partitionFrom; (partitionTo == -1 && partition == partitionFrom) || (partitionTo > 0 && partition < partitionTo); partition++ {
+		metricDefinitions = cassandraIndex.LoadPartitions([]int32{int32(partition)}, metricDefinitions, time.Now())
+	}
+
+	tracker := Tracker{}
+	for _, metricDefinition := range metricDefinitions {
+		tracker[metricDefinition.Id] = Track{
+			Name: metricDefinition.Name,
+			Tags: metricDefinition.Tags,
+		}
+	}
+
+	return &inputOOOFinder{
+		prefix: prefix,
+		substr: substr,
+
+		tracker:                  tracker,
+		reorderBufferGracePeriod: int64(metricInterval * reorderWindow),
+
+		groupByName:   groupByName,
+		groupedByName: groupedByName,
+		groupByTag:    groupByTag,
+		groupedByTag:  groupedByTag,
+
+		lock: sync.Mutex{},
+	}
+}
+
+func (ip *inputOOOFinder) processTrack(metricKey schema.MKey, metricTime int64, track Track) {
+	if ip.prefix != "" && !strings.HasPrefix(track.Name, ip.prefix) {
+		return
+	}
+	if ip.substr != "" && !strings.Contains(track.Name, ip.substr) {
+		return
+	}
+
+	if metricTime > track.LatestTimestamp {
+		track.LatestTimestamp = metricTime
+		ip.tracker[metricKey] = track
+	} else if metricTime+ip.reorderBufferGracePeriod < track.LatestTimestamp {
```

Comment:
There is another case here that isn't handled: duplicate data. If a timestamp is published twice, that should be counted separately from out of order data.